cylc/flow/network/scan.py

```diff
@@ -13,298 +13,361 @@
+over them using an ``async for`` statement::
+
+    async for flow in scan():
+        print(flow['name'])
+
+For further functionality construct a pipe::
+
+    pipe = scan | is_active(True) | contact_info
+    async for flow in pipe:
+        print(f'{flow["name"]} {flow["CYLC_SUITE_HOST"]}')
+
+There are filters which you can you to omit workflows e.g.
+:py:func:`cylc_version` and transformers which acquire more information
+e.g. :py:func:`contact_info`.
+
+.. note: we must manually list functions so they get built into the docs
+
+.. autofunction:: scan
+.. autofunction:: filter_name
+.. autofunction:: is_active
+.. autofunction:: contact_info
+.. autofunction:: cylc_version
+.. autofunction:: api_version
+.. autofunction:: graphql_query
+.. autofunction:: title
+
+"""
+
+from collections.abc import Iterable
 import asyncio
-import os
-from pwd import getpwall
+from pathlib import Path
 import re
-import sys
-import socket
-from cylc.flow import LOG
 
+from pkg_resources import (
+    parse_requirements,
+    parse_version
+)
+
+from cylc.flow import LOG
+from cylc.flow.async_util import (
+    pipe,
+    asyncqgen,
+    scandir
+)
 from cylc.flow.cfgspec.glbl_cfg import glbl_cfg
-from cylc.flow.exceptions import SuiteServiceFileError
-import cylc.flow.flags
-from cylc.flow.hostuserutil import is_remote_host, get_host_ip_by_name
 from cylc.flow.network.client import (
     SuiteRuntimeClient, ClientError, ClientTimeout)
+from cylc.flow.exceptions import SuiteStopped
 from cylc.flow.suite_files import (
     ContactFileFields,
     SuiteFiles,
-    load_contact_file,
     get_suite_title,
-    get_suite_source_dir
+    load_contact_file_async
 )
 
-DEBUG_DELIM = '\n' + ' ' * 4
-INACTIVITY_TIMEOUT = 10.0
-MSG_QUIT = "QUIT"
-MSG_TIMEOUT = "TIMEOUT"
-SLEEP_INTERVAL = 0.01
 
+SERVICE = Path(SuiteFiles.Service.DIRNAME)
+CONTACT = Path(SuiteFiles.Service.CONTACT)
+SUITERC = Path(SuiteFiles.SUITE_RC)
 
-def async_map(coroutine, iterator):
-    """Map iterator iterator onto a coroutine.
 
-    * Yields results in order as and when they are ready.
-    * Slow workers can block.
+async def dir_is_flow(listing):
+    """Return True if a Path contains a flow at the top level.
 
     Args:
-        coroutine (asyncio.coroutine):
-            I.E. an async function.
-        iterator (iter):
-            Should yield tuples to be passed into the coroutine.
+        listing (list):
+            A listing of the directory in question as a list of
+            ``pathlib.Path`` objects.
+
+    Returns:
+        bool - True if the listing indicates that this is a flow directory.
+
+    """
+    listing = {
+        path.name
+        for path in listing
+    }
+    return (
+        SERVICE.name in listing
+        or SUITERC.name in listing  # cylc7 flow definition file name
+        or 'flow.cylc' in listing  # cylc8 flow definition file name
+        or 'log' in listing
+    )
+
+
+@pipe
+async def scan(run_dir=None, scan_dir=None, max_depth=4):
```

Comment:
The `max_depth` is a new feature to avoid going too deep down the rabbit warren in the event that:

* Someone puts something which isn't a workflow into their cylc-run dir (a not entirely uncommon occurrence).
* Incase the way we recognise flows changes (e.g. the suite.rc file gets renamed or converted to a Python file).

What is a sensible default value for this?

Set as 4 for now as that's what's required by cylc functional test battery.

Reply:
I think probably too rare to care, or at least the default should be sufficient that we wouldn't expect users to hit it.